src/DataProtection/DataProtection/src/KeyManagement/KeyRingBasedDataProtector.cs

```diff
@@ -381,6 +353,118 @@ public byte[] GetAadForKey(Guid keyId, bool isProtecting)
             }
         }
 
+#if NET10_0_OR_GREATER
+        public AdditionalAuthenticatedDataTemplate(string[] purposes)
+        {
+            // additionalAuthenticatedData := { magicHeader (32-bit) || keyId || purposeCount (32-bit) || (purpose)* }
+            // purpose := { utf8ByteCount (7-bit encoded) || utf8Text }
+
+            var keySize = sizeof(Guid);
+            int totalPurposeLen = 4 + keySize + 4;
+
+            var purposeLengthsPool = ArrayPool<int>.Shared.Rent(purposes.Length);
+            for (int i = 0; i < purposes.Length; i++)
+            {
+                string purpose = purposes[i];
+
+                int purposeLength = EncodingUtil.SecureUtf8Encoding.GetByteCount(purpose);
+                purposeLengthsPool[i] = purposeLength;
+
+                var encoded7BitUIntLength = Measure7BitEncodedUIntLength((uint)purposeLength);
+                totalPurposeLen += purposeLength /* length of actual string */ + encoded7BitUIntLength /* length of 'string length' 7-bit encoded int */;
+            }
+
+            byte[] targetArr = new byte[totalPurposeLen];
+            var targetSpan = targetArr.AsSpan();
+
+            // index 0: magic header
+            BinaryPrimitives.WriteUInt32BigEndian(targetSpan.Slice(0), MAGIC_HEADER_V0);
+            // index 4: key (skipped for now, will be populated in `GetAadForKey()`)
+            // index 4 + keySize: purposeCount
+            BinaryPrimitives.WriteInt32BigEndian(targetSpan.Slice(4 + keySize), purposes.Length);
+
+            int index = 4 + keySize + 4; // starting from first purpose
+            for (int i = 0; i < purposes.Length; i++)
+            {
+                string purpose = purposes[i];
+
+                // writing `utf8ByteCount (7-bit encoded integer) || utf8Text`
+                // we have already calculated the lengths of the purpose strings, so just get it from the pool
+                index += Write7BitEncodedInt(purposeLengthsPool[i], targetSpan.Slice(index));
+                index += EncodingUtil.SecureUtf8Encoding.GetBytes(purpose.AsSpan(), targetSpan.Slice(index));
+            }
+
+            ArrayPool<int>.Shared.Return(purposeLengthsPool);
+            Debug.Assert(index == targetArr.Length);
+
+            _aadTemplate = targetArr;
+        }
+
+        private static int Measure7BitEncodedUIntLength(uint value)
+        {
+            return ((31 - System.Numerics.BitOperations.LeadingZeroCount(value | 1)) / 7) + 1;
```

Comment:
I know this is my code - however, I'd like to see a test on this specifically, in particular checking all the 7-bit boundaries, i.e. a `[Theory]` with tests for:

- 0 => 1
- 1 => 1
- 0b0_1111111 => 1
- 0b1_0000000 => 2
- 0b1111111_1111111 => 2
- 0b1_0000000_0000000 => 3
- 0b1111111_1111111_1111111 => 3
- 0b1_0000000_0000000_0000000 => 4
- 0b1111111_1111111_1111111_1111111 => 4
- 0b1_0000000_0000000_0000000_0000000 => 5
- uint.MaxValue =>5

Reply:
added `Measure7BitEncodedUIntLength_ReturnsExceptedLength()`